src/renderer/contexts/DialogContext.tsx

```diff
@@ -0,0 +1,91 @@
+import React, { createElement, useCallback, useMemo, useState } from 'react'
+
+import type { JSXElementConstructor, PropsWithChildren } from 'react'
+
+import { generateRandomUUID } from '../utils/random'
+
+type DialogId = string
+
+export type DialogProps = {
+  onClose: () => void
+}
+
+type DialogElementConstructor<T> = JSXElementConstructor<DialogProps & T>
+
+export type OpenDialog = <T extends { [key: string]: any }>(
+  dialogElement: DialogElementConstructor<T>,
+  additionalProps?: T
+) => DialogId
+
+export type CloseDialog = (id: DialogId) => void
+
+type DialogContextValue = {
+  hasOpenDialogs: boolean
+  openDialog: OpenDialog
+  closeDialog: CloseDialog
+}
+
+const initialValues: DialogContextValue = {
+  hasOpenDialogs: false,
+  openDialog: _ => '',
+  closeDialog: _ => {},
+}
+
+export const DialogContext = React.createContext<DialogContextValue>(
+  initialValues
+)
+
+export const DialogContextProvider = ({ children }: PropsWithChildren<{}>) => {
```

Comment:
This is the "heart" of this refactor PR: A new DialogContext and Provider. It is inspired by the previous implementation which solved a couple of problems:

* Have knowledge about if there's currently a dialogue in the view or not
* Have a common place where they are rendered
* Have a nice convenient `openDialog` API with type checks on the properties